sklearn/naive_bayes.py

```diff
@@ -949,3 +988,264 @@ def _joint_log_likelihood(self, X):
+            # update category_count_dimensions in case partial_fit is used
+            self.category_count_[i] = self._update_category_count_dimensions(
+                self.category_count_[i], self.feature_cat_index_mapping_[i])
+            self._update_category_count(X_feature, Y,
+                                        self.feature_cat_index_mapping_[i],
+                                        self.category_count_[i])
+
+    def _count_categories(self, X_feature):
+        if CategoricalNB.old_numpy_:
+            bins = np.bincount(X_feature)
+            non_zero_bins = np.nonzero(bins)[0]
+            return non_zero_bins, bins[non_zero_bins]
+        else:
+            return np.unique(X_feature, return_counts=True)
+
+    def _update_mapping(self, cat_mapping, categories):
+        for category in np.nditer(categories):
+            category = float(category)
+            if category not in cat_mapping:
+                cat_mapping[category] = len(cat_mapping)
+
+    def _update_category_count_dimensions(self, cat_count, cat_mapping):
+        diff = len(cat_mapping) - cat_count.shape[1]
+        if diff > 0:
+            b = np.zeros((self.class_count_.shape[0], len(cat_mapping)))
+            b[:, :cat_count.shape[1]] = cat_count
+            return b
+        return cat_count
+
+    def _update_category_count(self, X_feature, Y, cat_mapping, cat_count):
+        for j in range(self.class_count_.shape[0]):
+            X_feature_class = X_feature[Y[:, j] == 1]
+            if len(X_feature_class) == 0:
+                continue
+            class_cats, n_feature_class = \
+                self._count_categories(X_feature_class)
+            indices = [cat_mapping[float(category)] for category
+                       in np.nditer(class_cats)]
+            cat_count[j, indices] = n_feature_class
+
+    def _update_feature_log_prob(self, alpha):
+        feature_log_prob = {}
+        for i in range(self.n_features_):
+            smoothed_fc = self.category_count_[i] + alpha
+            smoothed_cc = smoothed_fc.sum(axis=1)
+            feature_log_prob[i] = (np.log(smoothed_fc) -
+                                   np.log(smoothed_cc.reshape(-1, 1)))
+        self.feature_log_prob_ = feature_log_prob
+
+    def _joint_log_likelihood(self, X):
+        check_is_fitted(self, "classes_")
+
+        if not X.shape[1] == self.n_features_:
+            raise ValueError(
+                'X has a different number of features than expected.')
+
+        samples_prob_zero = []
+        jll = np.zeros((X.shape[0], self.class_count_.shape[0]))
+        for i in range(self.n_features_):
+            X_feature = X[:, i]
+            indices = []
+            for sample, category in enumerate(np.nditer(X_feature)):
+                try:
+                    indices.append(
+                        self.feature_cat_index_mapping_[i][float(category)])
+                except KeyError:
+                    if self.on_unseen_cats == 'ignore':
```

Comment:
I wonder if it would be more comprehensible to replace this logic with:
```
if self.on_unseen_cats == 'warn':
    warnings.warn(...)
elif self.on_unseen_cats == 'raise':
    raise ...
```

Reply:
I see your point. The advantage of having an `else` statement at the end, though, is that if the user has a typo in his string, the program will raise an error.

Reply:
But you already check in the `__init__` that there is no typo, i.e.
```
if on_unseen_cats not in ('ignore', 'warn', 'raise'):
```
But anyway, at that point you should not handle misspelled settings...